content/blog/2020-12-18-db-compiler/index.md

```diff
@@ -0,0 +1,293 @@
+Imagine the following equivalent code:
+```
+query {
+  output = []
+  for (t in T) {
+    if (t[x] > 5) {
+      output.add(t)
+    }
+  }
+  return output
+}
+```
+This should be much faster since there are no function calls (let alone virtual ones) and since tight loops are generated, values are kept in registered for as long as possible.
+
+## High Level Design
+
+### Producer/Consumer Interface
+
+To generate code for an operator tree, we introduce produce and consume functions for each operator. Produce  generates a loop over the tuples in an operator's result. Consume is called by an operator's child surrounded by a loop over its result tuples. Effectively, produce signifies to "pull" tuples from an operator while consume "push"es them. Expressions have a produce function that generate code for calculating that expression.
+
+To see this in action, let's consider the following psuedocode implementations for Scan and Filter:
+```
+Scan.produce() {
+  output << "for (tuple in T) {"
+    parent.consume()
+  output << "}"
+}
+
+Filter.produce() {
+  // To generate a loop over its result, generate a loop over its child's result.
+  child.produce();
+}
+
+Filter.consume() {
+  // This is called in a loop over the child's result
+  output << "if (";
+    cond.produce()
+  output << ") {"
+    parent.consume()
+  output << "}"
+}
+```
+
+It's worth taking some time to appreciate how elegant this is and check that it does indeed do what we want. I was positively awestruck when I first came across it.
+
+Special care is needed for joins and group bys since they store some of the input into a hash table and then loop over it later. Here's some psuedocode for a hash-based join:
+
+```
+Join.produce() {
+  // First, generate a loop over one side
+  left_child.produce()
+  // See consume for what gets generated in that loop
+
+  // Then, generate a loop over the other side
+  right_child.produce()
+  // See consume for what gets generated in that loop
+}
+
+Join.consume(src) {
+  if (src == left_child) {
+    // This is called in a loop over the left child's result
+    output << "add t into HT"
+  } else {
+    // This is called in a loop over the right child's result
+    output << "for (t' in HT[t]) {"
+      parent.consume(this)
+    output << "}"
+  }
+}
+```
+What is needed to make this work is for us to specialize the consume function based on which loop this is generated in. We accomplish this by invoking the consume function with the operator that generated the loop. Similarly for group by:
+```
+GroupBy.produce() {
+  // First, generate a loop over child's result
+  child.produce()
+  // See consume for what gets generated in that loop
+
+  // Then, loop over each bucket and then output the aggregation functions
+  output << "for (bucket in HT) {"
+    output << "t = compute(bucket)"
+    parent.consume()
+  output << "}"
+}
+
+GroupBy.consume() {
+  output << "insert t into HT[group by cols of t]"
+}
+```
+
+## Implementation
```

Comment:
Maybe a couple of sentences of overview here would help clarify what you've actually built. Did you implement a whole DBMS? Just a query engine? If so, how does the query engine act on actual data? A little context here would be useful.